src/zmpl/Template.zig

```diff
@@ -409,33 +500,94 @@ fn compileMarkupLine(self: *Self, line: []const u8) ![]const u8 {
     return markup_line.compile();
 }
 
-fn compileRawLine(self: *Self, line: []const u8, chomp: bool) ![]const u8 {
-    const chomped = if (chomp) chompString(line) else line;
+fn compileRaw(self: *Self, string: []const u8, chomp: bool) ![]const u8 {
+    const chomped = if (chomp) chompString(string) else string;
+    const escaped_backslash = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        chomped,
+        "\"",
+        "\\\"",
+    );
+    defer self.allocator.free(escaped_backslash);
+
+    const escaped_linebreak_rn = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        escaped_backslash,
+        "\r\n",
+        "\\r\\n",
+    );
+    defer self.allocator.free(escaped_linebreak_rn);
+
+    const escaped_linebreak_n = try std.mem.replaceOwned(
+        u8,
+        self.allocator,
+        escaped_linebreak_rn,
+        "\n",
+        "\\n",
+    );
+    defer self.allocator.free(escaped_linebreak_n);
+
+    const compiled = escaped_linebreak_n;
+
     return std.mem.join(
         self.allocator,
         "",
         &[_][]const u8{
             "try zmpl.write(\"",
-            try std.mem.replaceOwned(u8, self.allocator, chomped, "\"", "\\\""),
+            compiled,
             "\\n\");",
         },
     );
 }
 
 fn compileZigLine(self: *Self, line: []const u8) ![]const u8 {
-    _ = self;
-    return line;
+    return try self.allocator.dupe(u8, line);
 }
 
 fn escapeText(self: *Self, text: []const u8) ![]const u8 {
     _ = self;
     return text;
 }
 
+fn escapeChar(char: u8) []const u8 {
+    if (char == '\r') return "\\r";
+    if (char == '\n') return "\\n";
+    if (char == '"') return "\"";
+    if (char == '\\') return "\\\\";
+
+    return &[_]u8{char};
+}
+
 fn chompString(string: []const u8) []const u8 {
     if (std.mem.endsWith(u8, string, "\r\n")) {
         return string[0 .. string.len - 3];
     } else if (std.mem.endsWith(u8, string, "\n")) {
         return string[0 .. string.len - 2];
     } else return string;
 }
+
+fn isWhitespace(string: []const u8) bool {
+    for (string) |char| {
+        if (char == ' ' or char == '\t') continue;
+        return false;
+    }
+    return true;
+}
+
+fn clearDanglingWhitespace(buf: *std.ArrayList(u8)) void {
+    var index: ?usize = null;
+
+    if (std.mem.lastIndexOf(u8, buf.items, "\r\n")) |linebreak_index| index = linebreak_index;
+    if (std.mem.lastIndexOf(u8, buf.items, "\n")) |linebreak_index| index = linebreak_index;
```

Comment:
I think this line makes more sense if swapped with the line above

Reply:
Oooops, I think I changed the implementation up, this was definitely deliberately ordered to make sure we catch `\r\n` first and don't accidentally leave a dangling `\r` - this is 100% a bug, will fix, thank you.